### ISSUE TITLE
Specify kind of vending machine

### ISSUE THREAD

**Issue by matkoniecz** [streetcomplete/StreetComplete#3587]
### General
Affected tag(s) to be modified/added: [vending](https://wiki.openstreetmap.org/wiki/Key:vending)

Question asked: **What is sold in this vending machine?**

### Checklist
Checklist for quest suggestions (see [guidelines](https://github.com/streetcomplete/StreetComplete/wiki/Adding-new-Quests-to-StreetComplete)):
- [x] 🚧 To be added tag is established and has a useful purpose
- [x] 🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
- [x] 🐿️ Easily answerable by everyone from the outside but a survey is necessary
- [x] 💤 Not an overwhelming percentage of quests have the same answer (No spam)
- [x] 🕓 Applies to a reasonable number of map data (Worth the effort) - around 8 000. Is it large enough? See https://taginfo.openstreetmap.org/tags/amenity=vending_machine#combinations 

From looking at http://overpass-turbo.eu/s/1e5K it appears that it is organic buildup, not some broken import or single user producing that.

![screen](https://user-images.githubusercontent.com/899988/145727350-7a2f3aa8-e14d-433c-8084-302594d94ecf.png)

### Ideas for implementation

**Element selection:** `amenity=vending_machine and !vending`

**Metadata needed:** Probably no?

**Proposed UI:** Image selection, from popular values: https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dvending_machine

**Comment by westnordost** [comment]
8000, that's pretty rare.

**Comment by matkoniecz** [comment]
Though on other hand this would rescue objects that are currently missing its main tag and are basically `fixme`.

I am considering trying it as it looks relatively simple and images of products should be relatively easy to find.

**Comment by coolultra1** [comment]
A big list of possible `vending` values. Somewhere, you'd have to draw the line between "This needs to be a quest option" and "This is just too rare, this should rather become an object, unfixable for SC users".

Also note that about 1/8 of the typeless vending machines have a (mostly descriptive) `name=` tag, which might help, should be cleaned up too though. I don't like leaving stuff dirty, even if it already was before.

**Comment by benpro** [comment]
In Japan, land of vending machines (more than 5 millions), which are sadly rarely mapped.
I am adding those from time to time with OsmAnd.
I think this quest is more than wanted in Japan, it will be not that much of a rare quest.

**Comment by kmpoppe** [comment]
> https://taginfo.openstreetmap.org/tags/amenity=vending_machine#combinations

<html><body>


Total Nr | Percentage | tag 
-- | -- | --
51 651 | 25.61% | vending=parking_tickets 
29 173 | 14.46% | vending=excrement_bags 
23 123 |11.46% | vending=cigarettes 
21 851 | 10.83% | vending=public_transport_tickets 
20 370 |10.10% | vending=drinks 
10 010| 4.96% | vending=parcel_pickup;parcel_mail_in 


</body>
</html>

I guess, apart from parcel_pickup, the other vending types that are the Top 5 would be a good starting point for them to be shown as a selector list?

K

**Comment by westnordost** [comment]
I still think this is not worth the effort. 

Some mapper who wanted to map an excrement bag "vending machine" will certainly not tag it simply as a "vending machine" without any extra tags. Same with parking tickets, cigarettes, public transport tickets and "DHL packstations" (parcel_pickup;parcel_mail_in).
So, the 8000 other vending machines are likely vending machines that are vending something else.

**Comment by InsertUser** [comment]
Is this worth doing now that StreetComplete _already has_ a bunch of icons for it via the things layer? It does include a generic one as the first option in the list which might be used by people who see one from a distance. 

At the moment it's only possible to convert from one type to another by deleting and recreating the node and that's not really good practice. It also doesn't let people select multiple values.

**Comment by matkoniecz** [comment]
https://github.com/streetcomplete/StreetComplete/issues/3587#issuecomment-1062285628 problem still mostly applies

https://overpass-turbo.eu/s/1RSv shows we have about 9500 objects worldwide which is also not really a lot

**Comment by mnalis** [comment]
**TL;DR:** maybe reopen as `needs PR` ?

---

> https://github.com/streetcomplete/StreetComplete/issues/3587#issuecomment-1062285628 problem still mostly applies

True, but I interpreted it mostly as suggestion why top-5 taginfo might not be a best choice for a final list of answers?

> https://overpass-turbo.eu/s/1RSv shows we have about 9500 objects worldwide which is also not really a lot

9500 objects is not that little either (and it is also noticeably more than 8000 where it was when quest was initially suggested) though.

> At the moment it's only possible to convert from one type to another by deleting and recreating the node and that's not really good practice

Yeah I'd find that as a horrible workaround

---

So the biggest complaint was _"probably not worth doing"_ part of that @westnordost comment, but that seems more like `needs PR` than outright rejection to me? (so self-close was maybe a little premature?)

IMHO **text-only list quest** (e.g. in format of `Is smoking allowed here?` quest) with top-10 answers (filtering out things unlikely to be tagged just as _"vending machines"_ like _excrement bags_, _parking tickets_ & _parcel pickups_) should be relatively **easy to do**. (and any notes created with it might show other popular values might be included in the list later) 

(I might write the code if that would remove the main objection of _"probably not worth doing"_?)

**Comment by matkoniecz** [comment]
> True, but I interpreted it mostly as suggestion why top-5 taginfo might not be a best choice for a final list of answers?

I interpreted this as maintenance cost associated with having such quest (especially given risk of bad answers given that some answers are likely to be missing)

**Comment by InsertUser** [comment]
The things layer appears to support 23 types. And an option for a freeform description would be reasonable is remaining cases.

>  (filtering out things unlikely to be tagged just as "vending machines" like excrement bags, parking tickets & parcel pickups) should be relatively easy to do

The ones near me are ones I suspect are parking ticket machines. They have mostly been added via OsmAnd that only makes it east to add a top level tag.

**Comment by mnalis** [comment]
> The things layer appears to support 23 types

Yeah, well, searching through the list of 23 types is probably going to fail _"🐿️ Easy answer"_, at least in StreetComplete and at least with that `smoking`-type quest UI. Even that top-10 answers was at a limit (top-5 would've  been more reasonable, but probably too little to cover even the most popular values) 

Searchable list (a'la Things Overlay) might be more acceptable for such amounts like 23 types, but that does not exist as a quest type AFAIK, so implementing that would be even more work.  (which in turn taxes _"🕓 Effort vs impact"_ even more, i.e. I wouldn't be the one volunteering to work on that, as it is beyond my capabilities...)

> And an option for a freeform description would be reasonable is remaining cases.

There is `Uh...` / `Can't say` / `Leave note`, which is default _"freeform description of missing answer"_. Or was your intention something else?
 
> The ones near me are ones I suspect are parking ticket machines. They have mostly been added via OsmAnd that only makes it east to add a top level tag.

That is quite useful information esp. given OsmAnd popularity; so perhaps then even those "strange" top values should be retained in the list of the answers (assuming the quest idea is still salvageable)...

**Comment by InsertUser** [comment]
Looking at the quest list again it appears that the "What sport is played here?" quest has `;` separated multivalue lists _and_ 33 options to choose from, which would be plenty. Allowing multiple values would be useful for things like `drinks;snacks` which I have seen quite often in the real world even if it isn't common in the OSM database.
 
The downside of that is that it emphasises icons and the icons we already have are quite similar (and some are repeats). If the license permits we could potentially borrow some from the [equivalent MapComplete theme](https://mapcomplete.org/vending_machine.html?test=true).

I was toying with the idea of sticking a freeform value in the description tag, as I think already happens with a generic "unpaved" surface in StreetComplete, but maybe leaving a note for a more experienced mapper is the best option.

**Comment by mnalis** [comment]
> The downside of that is that it emphasises icons and the icons we already have are quite similar (and some are repeats). If the license permits we could potentially borrow some from the [equivalent MapComplete theme](https://mapcomplete.org/vending_machine.html?test=true).

Yeah, the icon quest would work better, but it requires _significantly_ more effort than text-only quest. MapComplete black/white simple icons wouldn't cut it for StreetComplete, you'd need more detailed and colorful ones, like the ones for `sport=*` quest.

But then _"pitch without sport"_ has over 312k unsolved quests, and this only has 9k, so _"🕓 Effort vs impact"_ is 35 times less favorable here :cry:. Now, if someone were to invest effort to collect (and create missing) those 20-30 nice color SVG icons for vending machine products, that would be great - and quest could easily be written.

But unless someone actually volunteers to do that effort, that quest currently doesn't seem too likely to happen to me... :man_shrugging: 

> but maybe leaving a note for a more experienced mapper is the best option

Yeah, I think OSM Note would be best for those remaining little-used values (if the quest were to happen)...

**Comment by InsertUser** [comment]
The amount of effort required for the colourful emojis is a big downside to that approach. 

Some of the more colourful ones on the MapComplete theme appear to be from openclipart.org which says it's CC0 1.0. A quick search of that shows a couple that look a bit iffy WRT e.g. recognisable trademarks, but I might have a bit of a look through there and Wikimedia Commons and see If I can get started on a list that don't look too far out of place. The ["squished coin" one](https://mapcomplete.org/assets/themes/elongated_coin/penny.svg) looks fine as is to me, so that's one down, one score to go. 

I don't think GitHub allows you to mark a post as a "community wiki" as some other sites so if I do this I'll probably do it under my user space on the OSM wiki and link to it from here. That way If I fail and someone else rediscovers this request in a couple of years they can continue to add to it.

**Comment by westnordost** [comment]
svgrepo is usually a good source, they aggregate icons from many different sources.

**Comment by InsertUser** [comment]
OK, So I've gone through what the most popular values were when I last posted to this thread and added a bunch of icons to a table and placed it [here](https://wiki.openstreetmap.org/wiki/User:InsertUser/StreetComplete_Vending_Machine_Refinement). I've got most of the first couple dozen and most of the ones that street complete can add by other means. 

As it transpires, converting between text based markup languages is not as fun as it sounds. 

I'm not sure how far down the list we'd actually want to go with the quest. I have indicated whether a vending type is already selectable in the things layer for new objects. 

While I think my Inkscape knowledge might be up to making new icons, my artistic ability probably isn't. If there is a gap that we think absolutely needs to be covered I might have a go though. 

Which of these would we actually want to include?

**Comment by westnordost** [comment]
Wow, a lot of work. And interesting that so many vending machines are already available via the SC things overlay, i.e. are in the iD editor presets. Does that mean that all those vending machines already have each a different icon in iD presets? Just checked, indeed, _some_ do.

Given that the amount of things that can be vended in vending machines is open-ended and how well-spread the usage is, I think your table shows that attempting to create a UI like for the recycling or sports quest with a neat icon for each would not only be a Sisyphean task, it would actually not be convenient for users, either.

So, another possible TBD UI could look like this:
- input field where user inputs what is sold
- while typing, all iD editor presets that match what is typed are shown in a dropdown (aka popup menu). **Search results are limited to vending machines only.**
- user taps an item in the dropdown to apply it

I.e. basically the same UI as for the things overlay or the "It was empty, what's here now" dialog, with the modification that only search results matching a given set of tags (in this case: `amenity=vending_machine` and maybe a few more) are shown.

This is the same UI I envision would make sense for #1020. So, as far as working towards creating such a quest (without actually implementing it yet) goes, one could contribute presets to the iD editor presets repo.

**Comment by InsertUser** [comment]
I think multi-select should be possible in whatever interface is used. For example `=food` may have ~4k uses in the database, but various combinations of `food`, `drinks` and `sweets` expands that by about 2k. That does apply to healthcare too though where clinics might have several doctors with different specialties. 

I do like the idea of "search and add to list" as a way of doing this.

**Comment by tiptoptom** [comment]
There is already a similar quest form in SCEE for destinations. In this case, the suggestions are generated from the last entries.
https://github.com/Helium314/SCEE

![Image](https://github.com/user-attachments/assets/12f35b9d-099b-4d1b-a236-b7ca0844b67d)

**Comment by mnalis** [comment]
> There is already a similar quest form in SCEE for destinations

Just to note, if one is using [SCEE](https://wiki.openstreetmap.org/wiki/SCEE), they can already (after enabling _"Expert mode"_ in settings):
- have pseudo-quest showing all `amenity=vending_machine` without `vending=*` (by modifying `POI:machine` "half-quest")
- ability to select them and add `vending=*` from list of popular presets (or enter custom one)

See e.g. this video:

https://github.com/user-attachments/assets/a5cec7d7-ee3e-4c02-9ad1-870d2b0f5705

**Comment by westnordost** [comment]
If I followed the discussion correctly, my main reservation against this has been "probably not worth the implementation effort", so I'll reopen as "needs pr".

As I've mentioned earlier, this quest would use the same TBD UI as #1020, i.e. a variant of the iD-presets-based feature-search dialog with a custom filter applied and likely also a custom list of suggestions (i.e. what would be supermarket, café, restaurant etc. in the places search dialog) which would just be the complete list of - in this case - vending machine types.

Or in other words, it would work quite similarly as the quest selection screen - a list of all possible options is displayed and above it, a search field with which the results can be curtailed.
